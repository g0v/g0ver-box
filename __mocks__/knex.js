import knex from 'jest-mock-knex';

export { client } from 'jest-mock-knex';

export default knex;

process.setMaxListeners(0);
