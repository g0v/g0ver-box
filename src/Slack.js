import _ from 'lodash';
import slack from 'slack';

export const bot = slack.rtm.client();

const TOKEN = process.env.SLACK_TOKEN;

if (TOKEN) {
  bot.listen({ token: TOKEN });
}

bot.hello((message) => {
  console.log(`Got a message: ${message.type}`);
});

const call = (name, req = {}) => (
  new Promise((resolve, reject) => {
    const fn = _.get(slack, name);
    fn(
      { ...req, token: TOKEN },
      (err, res) => (err ? reject(err) : resolve(res))
    );
  })
);

export default {
  userInfo: async req => await call('users.info', req),
  postMessage: async req => await call('chat.postMessage', { ...req, as_user: true }),
  postMultiMessage: async (users, message) => _.map(users, user => bot.ws.send(JSON.stringify({
    id: `${user}::${Date.now()}`, type: 'message', channel: user, text: message,
  }))),
  channelInfo: async req => await call('channels.info', req),
  channelJoin: async req => await call('channels.join', req),
};
