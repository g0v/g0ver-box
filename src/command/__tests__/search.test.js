import { client } from 'knex';
import Slack from '../../Slack';
import index from '../';

describe('search command', () => {
  const data = { channel: 'D100', user: 'U03B2AB13', name: 'yutin', text: 'search react' };

  it('when g0ver not find', async () => {
    client.mockReturnValueOnce([]);
    await index(data);
    expect(client).toMatchSnapshot();
    expect(Slack.postMessage).toHaveBeenLastCalledWith({ channel: 'D100', text: '搜尋 react 找不到 g0ver, 也許你就是這樣沒有人.' });
  });

  it('when g0ver is existed', async () => {
    client.mockReturnValueOnce([
      { id: 'U03B2AB13', skills: ['react', 'video'] },
      { id: 'U0RQYV16K', skills: ['react', 'math'] },
    ]);
    await index(data);
    expect(client).toMatchSnapshot();
    expect(Slack.postMessage).toHaveBeenLastCalledWith({ channel: 'D100', text: '搜尋 react 找到了這些 g0ver: <@U03B2AB13>, <@U0RQYV16K>.' });
  });
});
