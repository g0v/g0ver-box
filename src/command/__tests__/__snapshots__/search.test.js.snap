// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`search command when g0ver is existed 1`] = `
Array [
  "select *, ts_rank(keyword, react) as rank from g0ver where keyword @@ to_tsquery(react) and deleted_at is null order by rank desc limit 1000",
]
`;

exports[`search command when g0ver not find 1`] = `
Array [
  "select *, ts_rank(keyword, react) as rank from g0ver where keyword @@ to_tsquery(react) and deleted_at is null order by rank desc limit 1000",
]
`;
